component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,178 @@
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.ISSUER;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_ASSERTION;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_ASSERTION_TYPE;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.OAUTH_JWT_BEARER_GRANT_TYPE;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.PREVENT_TOKEN_REUSE;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.REJECT_BEFORE_IN_MINUTES;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.SIGNED_JWT;
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants.TOKEN_ENDPOINT_ALIAS;
+
+/**
+ * Client Authentication handler to implement oidc private_key_jwt client authentication spec
+ * http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.
+ */
+public class PrivateKeyJWTClientAuthenticator extends AbstractOAuthClientAuthenticator {
+
+    private static final Log log = LogFactory.getLog(PrivateKeyJWTClientAuthenticator.class);
+    private JWTValidator jwtValidator;
+
+    public PrivateKeyJWTClientAuthenticator() {
+
+        String tokenEPAlias = properties.getProperty(TOKEN_ENDPOINT_ALIAS);
+        String issuer = properties.getProperty(ISSUER);
+        boolean preventTokenReuse = Boolean.parseBoolean(properties.getProperty(PREVENT_TOKEN_REUSE));
+        int rejectBeforePeriod = DEFAULT_VALIDITY_PERIOD_IN_MINUTES;
+        if (properties.getProperty(REJECT_BEFORE_IN_MINUTES) != null) {
+            rejectBeforePeriod = Integer.parseInt(properties.getProperty(REJECT_BEFORE_IN_MINUTES));
+        }
+        jwtValidator = createJWTValidator(tokenEPAlias, issuer, preventTokenReuse, rejectBeforePeriod);
+    }
+
+    /**
+     * To check whether the authentication is successful.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param bodyParameters          map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the authentication is successful.
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public boolean authenticateClient(HttpServletRequest httpServletRequest, Map<String, List> bodyParameters,
+                                      OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        String oauthJWTAssertion = getBodyParameters(bodyParameters).get(OAUTH_JWT_ASSERTION);
+        Object signedJWT = oAuthClientAuthnContext.getParameter(SIGNED_JWT);
+        if (signedJWT != null) {
+            return jwtValidator.isValidAssertion((SignedJWT) signedJWT);
+        }
+        return jwtValidator.isValidAssertion(getSignedJWT(oauthJWTAssertion));
```

Comment:
Can refactor to move 
..
+ String oauthJWTAssertion = getBodyParameters(bodyParameters).get(OAUTH_JWT_ASSERTION);
 +        Object signedJWT = oAuthClientAuthnContext.getParameter(SIGNED_JWT);
 +        if (signedJWT != null) {
..
part in to getSignedJWT method.